src/esp/sim/ReplayBatchRenderer.h

```diff
@@ -23,75 +22,211 @@ class Renderer;
 
- private:
-  scene::SceneNode* loadAndCreateRenderAssetInstance(
-      int envIndex,
+  // void doSetEnvironmentKeyframe(int envIndex, const std::string& serKeyframe)
+  // override;
+
+  void doSetSensorTransform(unsigned envIndex,
+                            const std::string& sensorName,
+                            const Mn::Matrix4& transform) override;
+
+  void doSetSensorTransformsFromKeyframe(unsigned envIndex,
+                                         const std::string& prefix) override;
+
+  void doRender(Corrade::Containers::ArrayView<const Magnum::MutableImageView2D>
+                    imageViews) override;
+
+  void doRender(Magnum::GL::AbstractFramebuffer& framebuffer) override;
+
+  gfx::replay::NodeHandle loadAndCreateRenderAssetInstance(
+      unsigned envIndex,
       const assets::AssetInfo& assetInfo,
       const assets::RenderAssetInstanceCreationInfo& creation);
 
   std::vector<EnvironmentRecord> envs_;
 
-  std::unique_ptr<assets::ResourceManager> resourceManager_ = nullptr;
+  std::unique_ptr<assets::ResourceManager> resourceManager_;
 
   scene::SceneManager::uptr sceneManager_ = nullptr;
 
   gfx::WindowlessContext::uptr context_ = nullptr;
   std::shared_ptr<gfx::Renderer> renderer_ = nullptr;
 
-  ReplayBatchRendererConfiguration config_;
+  ReplayRendererConfiguration config_;
+
+  ESP_SMART_POINTERS(ReplayRenderer)
+};
+
+class ReplayBatchRenderer : public AbstractReplayRenderer {
+ public:
+  explicit ReplayBatchRenderer(const ReplayRendererConfiguration& cfg);
+
+  ~ReplayBatchRenderer() override;
+
+ private:
+  unsigned doEnvironmentCount() const override;
+
+  Magnum::Vector2i doSensorSize(unsigned envIndex) override;
+
+  esp::gfx::replay::Player& doPlayerFor(unsigned envIndex) override;
+
+  // void doSetEnvironmentKeyframe(unsigned envIndex, const std::string&
+  // serKeyframe) override;
+
+  void doSetSensorTransform(unsigned envIndex,
+                            const std::string& sensorName,
+                            const Mn::Matrix4& transform) override;
+
+  void doSetSensorTransformsFromKeyframe(unsigned envIndex,
+                                         const std::string& prefix) override;
+
+  void doRender(Corrade::Containers::ArrayView<const Magnum::MutableImageView2D>
+                    imageViews) override;
+
+  void doRender(Magnum::GL::AbstractFramebuffer& framebuffer) override;
+
+  // TODO pimpl all this?
+  struct EnvironmentRecord {
+    Corrade::Containers::Pointer<gfx::replay::AbstractPlayerImplementation>
+        playerImplementation_;
+    gfx::replay::Player player_{*playerImplementation_};
+  };
+  Corrade::Containers::Array<EnvironmentRecord> envs_;
+
+  /* If standalone_ is true, renderer_ contains a RendererStandalone */
+  bool standalone_;
+  Corrade::Containers::Pointer<esp::gfx_batch::Renderer> renderer_;
 
-  ESP_SMART_POINTERS(ReplayBatchRenderer)
+  Corrade::Containers::String theOnlySensorName_;
+  Mn::Matrix4 theOnlySensorProjection_;
 };
```

Comment:
`ESP_SMART_POINTERS(ReplayBatchRenderer)` is missing here.

Reply:
Yeah, I was wondering what do we actually need it for, and tried not adding it here :) Especially given this class would be used only through the base interface.

Reply:
Added.